### PR TITLE
feat: support pointer gesture

### PR DESCRIPTION
**Pull request by Groveer**
Add PointerGestureV1 QML type
Connect WSeat with PointerGestureV1

Log:

### CODE REVIEW

File: src/server/kernel/wseat.cpp [hunk 1]
```diff
@@ -960,6 +965,134 @@ void WSeat::notifyFrame(WCursor *cursor)
...
+                    scale, local, global, d->gestureFingers);
+    }
+
+    Q_EMIT sendPinchUpdate(time_msec, pos, scale, rotation);
+}
+
+void WSeat::notifyPinchEnd(WCursor *cursor, WInputDevice *device, uint32_t time_msec, bool cancelled)
+{
+    W_D(WSeat);
+    if (!d->gestureState) {
```
Comment (wineee) [inline src/server/kernel/wseat.cpp:1070]: qCWarning

File: src/server/kernel/wseat.cpp [hunk 2]
```diff
@@ -960,6 +965,134 @@ void WSeat::notifyFrame(WCursor *cursor)
...
+    QWindowSystemInterface::handleGestureEvent(w, time_msec, qwDevice,
+        Qt::NativeGestureType::BeginNativeGesture, local, global, fingers);
+
+    Q_EMIT sendPinchBegin(time_msec, fingers);
+}
+
+void WSeat::notifyPinchUpdate(WCursor *cursor, WInputDevice *device, uint32_t time_msec, const QPointF &pos, double scale, double rotation)
+{
+    W_D(WSeat);
+    if (!d->gestureState) {
```
Comment (wineee) [inline src/server/kernel/wseat.cpp:1041]: qCWarning

File: src/server/kernel/wseat.cpp [hunk 3]
```diff
@@ -960,6 +965,134 @@ void WSeat::notifyFrame(WCursor *cursor)
...
+        Qt::NativeGestureType::EndNativeGesture, local, global, d->gestureFingers);
+
+    Q_EMIT sendSwipeEnd(time_msec, cancelled);
+}
+
+void WSeat::notifyPinchBegin(WCursor *cursor, WInputDevice *device, uint32_t time_msec, uint32_t fingers)
+{
+    W_D(WSeat);
+    if (d->gestureState) {
+        qCWarning(qLcWlrGestureEvents) << "Unexpected GestureStarted while already active";
```
Comment (wineee) [inline src/server/kernel/wseat.cpp:1024]: return ?

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/kernel/wseat.cpp (hunk 3 above)
Comment (Groveer) [reply src/server/kernel/wseat.cpp:1024]: no retrun, allow rebegin event

**Review by wineee — APPROVED** [review]

File: src/server/kernel/wseat.cpp [hunk 4]
```diff
@@ -960,6 +965,138 @@ void WSeat::notifyFrame(WCursor *cursor)
...
+    auto *w = cursor->eventWindow();
+    const QPointF &global = cursor->position();
+    const QPointF local = w ? global - QPointF(w->position()) : QPointF();
+    QWindowSystemInterface::handleGestureEvent(w, time_msec, qwDevice,
+        Qt::NativeGestureType::BeginNativeGesture, local, global, fingers);
+
+    Q_EMIT sendSwipeBegin(time_msec, fingers);
+}
+
+void WSeat::notifySwipeUpdate(WCursor *cursor, WInputDevice *device, uint32_t time_msec, const QPointF &pos)
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:986]: `pos` -> `delta`

File: src/server/kernel/wseat.cpp [hunk 5]
```diff
@@ -960,6 +965,138 @@ void WSeat::notifyFrame(WCursor *cursor)
...
+    W_D(WSeat);
+    if (!d->gestureState) {
+        qCWarning(qLcWlrGestureEvents) << "Unexpected SwipeUpdate while not begin";
+        return;
+    }
+    auto qwDevice = qobject_cast<QPointingDevice*>(device->qtDevice());
+    auto *w = cursor->eventWindow();
+    const QPointF &global = cursor->position();
+    const QPointF local = w ? global - QPointF(w->position()) : QPointF();
+    QWindowSystemInterface::handleGestureEventWithValueAndDelta(w, time_msec, qwDevice,
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:997]: Why is not `SwipeNativeGesture`?

File: src/server/kernel/wseat.cpp [hunk 6]
```diff
@@ -960,6 +965,138 @@ void WSeat::notifyFrame(WCursor *cursor)
...
+    auto *w = cursor->eventWindow();
+    const QPointF &global = cursor->position();
+    const QPointF local = w ? global - QPointF(w->position()) : QPointF();
+    QWindowSystemInterface::handleGestureEvent(w, time_msec, qwDevice,
+        Qt::NativeGestureType::BeginNativeGesture, local, global, fingers);
+
+    Q_EMIT sendPinchBegin(time_msec, fingers);
+}
+
+void WSeat::notifyPinchUpdate(WCursor *cursor, WInputDevice *device, uint32_t time_msec, const QPointF &pos, double scale, double rotation)
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:1040]: `pos` -> `delta`

File: src/server/kernel/wseat.cpp [hunk 7]
```diff
@@ -960,6 +965,138 @@ void WSeat::notifyFrame(WCursor *cursor)
...
+    const QPointF &global = cursor->position();
+    const QPointF local = w ? global - QPointF(w->position()) : QPointF();
+
+    QWindowSystemInterface::handleGestureEvent(w, time_msec, qwDevice,
+        Qt::NativeGestureType::EndNativeGesture, local, global, d->gestureFingers);
+
+    Q_EMIT sendSwipeEnd(time_msec, cancelled);
+}
+
+void WSeat::notifyPinchBegin(WCursor *cursor, WInputDevice *device, uint32_t time_msec, uint32_t fingers)
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:1022]: Like as `notifySwipeBegin`? Can split a common function?

File: src/server/kernel/wseat.cpp [hunk 8]
```diff
@@ -960,6 +965,138 @@ void WSeat::notifyFrame(WCursor *cursor)
...
+    if (scale != 0) {
+        QWindowSystemInterface::handleGestureEventWithRealValue(
+                    w, time_msec, qwDevice, Qt::ZoomNativeGesture,
+                    scale, local, global, d->gestureFingers);
+    }
+
+    Q_EMIT sendPinchUpdate(time_msec, pos, scale, rotation);
+}
+
+void WSeat::notifyPinchEnd(WCursor *cursor, WInputDevice *device, uint32_t time_msec, bool cancelled)
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:1070]: Like as notifySwipeEnd? Can split a common function?

File: src/server/kernel/wseat.cpp [hunk 9]
```diff
@@ -960,6 +965,138 @@ void WSeat::notifyFrame(WCursor *cursor)
...
+
+    QWindowSystemInterface::handleGestureEvent(w, time_msec, qwDevice,
+        Qt::NativeGestureType::EndNativeGesture, local, global, d->gestureFingers);
+
+    Q_EMIT sendPinchEnd(time_msec, cancelled);
+}
+
+void WSeat::notifyHoldBegin(WCursor *cursor, WInputDevice *device, uint32_t time_msec, uint32_t fingers)
+{
+    // TODO: set qt event here.
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:1091]: Should send a touch event with `QEventPoint::State::Pressed` ?

File: src/server/qtquick/private/wquickpointergesturesv1.cpp [hunk 10]
```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2023 groveer <guoyao@uniontech.com>.
```
Comment (zccrs) [inline src/server/qtquick/private/wquickpointergesturesv1.cpp:1]: 2024

File: src/server/qtquick/private/wquickpointergesturesv1.cpp [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    WQuickWaylandServerInterface::create();
+    d->manager = QWPointerGesturesV1::create(server()->handle());
+}
+
+void WQuickPointerGesturesV1::setSeat(WSeat *seat)
+{
+    W_D(WQuickPointerGesturesV1);
+    if (d->seat != seat) {
+        if (d->seat != nullptr) {
+            disconnect(d->seat, &WSeat::sendSwipeBegin, this, &WQuickPointerGesturesV1::sendSwipeBegin);
```
Comment (zccrs) [inline src/server/qtquick/private/wquickpointergesturesv1.cpp:55]: No, all event of client's surface should from WSurfaceItem.

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/kernel/wseat.cpp (hunk 5 above)
Comment (Groveer) [reply src/server/kernel/wseat.cpp:997]: see 
https://github.com/qt/qtwayland/blob/a5bea434aa9137f9b0ba6170fbb5a237e890b021/src/client/qwaylandwindow.cpp#L1268

Comment (zccrs) [reply src/server/kernel/wseat.cpp:997]: Maybe is a Qt bug.

File: src/server/kernel/wseat.cpp [hunk 12]
```diff
@@ -960,6 +1000,78 @@ void WSeat::notifyFrame(WCursor *cursor)
...
+    }
+    d->gestureActive = true;
+    d->gestureFingers = fingers;
+    d->gestureType = gestureType;
+    auto qwDevice = qobject_cast<QPointingDevice*>(device->qtDevice());
+    auto *w = cursor->eventWindow();
+    const QPointF &global = cursor->position();
+    const QPointF local = w ? global - QPointF(w->position()) : QPointF();
+    QWindowSystemInterface::handleGestureEvent(w, time_msec, qwDevice,
+        Qt::NativeGestureType::BeginNativeGesture, local, global, fingers);
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:1017]: ???

File: src/server/kernel/wseat.cpp (hunk 9 above)
Comment (zccrs) [reply src/server/kernel/wseat.cpp:1068]: Use a custom event send to window. hold events must send to wayland clients.

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/kernel/wseat.cpp [hunk 13]
```diff
@@ -770,6 +797,46 @@ bool WSeat::sendEvent(WSurface *target, QObject *shellObject, QObject *eventObje
         d->doTouchNotifyCancel(target->handle());
         break;
     }
+    case QEvent::NativeGesture: {
```
Comment (wineee) [inline src/server/kernel/wseat.cpp:800]: if (d->gesture)  {
// sendEvent
}

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/qtquick/wsurfaceitem.cpp [hunk 14]
```diff
@@ -166,6 +166,14 @@ class EventItem : public QQuickItem
                 return true;
             break;
         }
+        case NativeGesture: {
+            auto e = static_cast<QNativeGestureEvent*>(event);
+            Q_ASSERT(e);
+            // We may receive HoverLeave when WSurfaceItem was destroying
+            if (parent() && static_cast<WSurfaceItem*>(parent())->sendEvent(e))
```
Comment (wineee) [inline src/server/qtquick/wsurfaceitem.cpp:173]: NativeGesture is not HoverLeave

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/kernel/wseat.cpp [hunk 15]
```diff
@@ -306,6 +327,7 @@ class WSeatPrivate : public WObjectPrivate
 
     QString name;
     WCursor* cursor = nullptr;
+    QWPointerGesturesV1* gesture = nullptr;
```
Comment (wineee) [inline src/server/kernel/wseat.cpp:330]:
```suggestion\
    WCursor *cursor = nullptr;
    QWPointerGesturesV1 *gesture = nullptr;
```

File: src/server/kernel/wseat.cpp [hunk 16]
```diff
@@ -71,6 +73,25 @@ class WSeatWheelEvent : public QWheelEvent {
 };
 #endif
 
+class WSeatHoldEvent : public QInputEvent {
+public:
+    static const QEvent::Type HoldBegin = static_cast<QEvent::Type>(2000);
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:78]: Use `QEvent::registerEventType`

File: src/server/kernel/wseat.cpp [hunk 17]
```diff
@@ -71,6 +73,25 @@ class WSeatWheelEvent : public QWheelEvent {
...
+    static const QEvent::Type HoldEnd = static_cast<QEvent::Type>(2001);
+    explicit WSeatHoldEvent(QEvent::Type type, const QPointingDevice *dev, int fingers = 1)
+        : QInputEvent(HoldBegin, dev)
+        , m_fingers(fingers)
+    {
+    }
+
+    inline int fingerCount() const { return m_fingers; }
+    inline void enableCancelled() { m_cancelled = true; }
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:87]: `setCancelled(bool)`

File: src/server/kernel/wseat.cpp (hunk 15 above)
Comment (zccrs) [inline src/server/kernel/wseat.cpp:330]: code style: `Type *`

File: src/server/kernel/wseat.cpp [hunk 18]
```diff
@@ -314,6 +336,11 @@ class WSeatPrivate : public WObjectPrivate
     QMetaObject::Connection onEventObjectDestroy;
     wlr_surface *oldPointerFocusSurface = nullptr;
 
+    bool gestureActive = false;
+    int gestureFingers = 0;
+    qreal lastScale = 1.0;
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:341]: Nees `lastRotation`?

File: src/server/qtquick/wsurfaceitem.cpp [hunk 19]
```diff
@@ -166,6 +166,14 @@ class EventItem : public QQuickItem
                 return true;
             break;
         }
+        case NativeGesture: {
+            auto e = static_cast<QNativeGestureEvent*>(event);
```
Comment (zccrs) [inline src/server/qtquick/wsurfaceitem.cpp:170]: Where process for `HoldBegin/HoldEnd` event?

**Review by zccrs — CHANGES_REQUESTED** [review]

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/kernel/wseat.cpp [hunk 20]
```diff
@@ -71,6 +73,25 @@ class WSeatWheelEvent : public QWheelEvent {
...
+        , m_fingers(fingers)
+    {
+    }
+
+    inline int fingerCount() const { return m_fingers; }
+    inline void enableCancelled() { m_cancelled = true; }
+    inline bool cancelled() const { return m_cancelled; }
+
+private:
+    int m_fingers = 0;
```
Comment (wineee) [inline src/server/kernel/wseat.cpp:91]: m_fingers default value should same in constructor.
`explicit WSeatHoldEvent(QEvent::Type type, const QPointingDevice *dev, int fingers = 1)`

File: src/server/kernel/wseat.cpp (hunk 17 above)
Comment (wineee) [reply src/server/kernel/wseat.cpp:87]: why not set  `m_cancelled` in constructor

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/kernel/wseat.cpp (hunk 18 above)
Comment (Groveer) [reply src/server/kernel/wseat.cpp:341]: not need

File: src/server/kernel/wcursor.cpp [hunk 21]
```diff
@@ -461,6 +537,23 @@ void WCursorPrivate::connect()
         on_frame();
     });
 
+    QObject::connect(handle, SIGNAL(swipeBegin(wlr_pointer_swipe_begin_event*)),
+                     q_func(), SLOT(on_swipe_begin(wlr_pointer_swipe_begin_event*)));
```
Comment (zccrs) [inline src/server/kernel/wcursor.cpp:541]: Use `W_Q` macro

File: src/server/kernel/wevent.h [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "wglobal.h"
+
+#include <QEvent>
+#include <QInputEvent>
+#include <QNativeGestureEvent>
+
+WAYLIB_SERVER_BEGIN_NAMESPACE
+
+class Q_GUI_EXPORT WGestureEvent : public QNativeGestureEvent {
```
Comment (zccrs) [inline src/server/kernel/wevent.h:12]: WAYLIB_SERVER_EXPORT

File: src/server/kernel/wseat.h [hunk 23]
```diff
@@ -3,6 +3,8 @@
 
 #pragma once
 
+#include "wevent.h"
```
Comment (zccrs) [inline src/server/kernel/wseat.h:6]: `<WEvent>`

File: src/server/kernel/wseat.h [hunk 24]
```diff
@@ -118,6 +120,13 @@ class WSeat : public WServerInterface, public WObject
                     double delta, int32_t delta_discrete, uint32_t timestamp);
     void notifyFrame(WCursor *cursor);
 
+    // swipe
```
Comment (zccrs) [inline src/server/kernel/wseat.h:123]: `// gesture`

File: src/server/qtquick/woutputrenderwindow.cpp [hunk 25]
```diff
@@ -1117,7 +1117,7 @@ WOutputRenderWindow::WOutputRenderWindow(QObject *parent)
     connect(contentItem(), &QQuickItem::widthChanged, this, &WOutputRenderWindow::widthChanged);
     connect(contentItem(), &QQuickItem::heightChanged, this, &WOutputRenderWindow::heightChanged);
     // renderwindow inherits qquickwindow, default no focusscope-isolation & no focus on contentItem(QQuickRootItem) when startup
-    // setFocus(true) to deliver focus to children on startup, 
+    // setFocus(true) to deliver focus to children on startup,
```
Comment (zccrs) [inline src/server/qtquick/woutputrenderwindow.cpp:1120]: ???

**Review by zccrs — CHANGES_REQUESTED** [review]

**Review by zccrs — APPROVED** [review]

File: src/server/kernel/wseat.cpp [hunk 26]
```diff
@@ -314,6 +317,11 @@ class WSeatPrivate : public WObjectPrivate
     QMetaObject::Connection onEventObjectDestroy;
     wlr_surface *oldPointerFocusSurface = nullptr;
 
+    bool disableGesture = false;
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:320]: No

File: src/server/kernel/wseat.cpp [hunk 27]
```diff
@@ -1113,6 +1269,9 @@ void WSeat::create(WServer *server)
         if (d->cursor)
             d->cursor->attachInputDevice(i);
     }
+    qgetenv("DISABLE_GESTURE").toInt(&d->disableGesture);
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:1272]: `qEnvironmentVariableIsSet("WAYLIB_DISABLE_GESTURE")`

File: src/server/kernel/wseat.cpp [hunk 28]
```diff
@@ -1113,6 +1269,9 @@ void WSeat::create(WServer *server)
...
+    if (!d->disableGesture)
```
Comment (zccrs) [inline src/server/kernel/wseat.cpp:1273]: if (!qEnvironmentVariableIsSet("WAYLIB_DISABLE_GESTURE"))

**Review by zccrs — CHANGES_REQUESTED** [review]

**Review by zccrs — APPROVED** [review]

**Pull request merged**